mq/inmemory/mq.h

```diff
@@ -2,30 +2,104 @@
+//      message, MMQ DOES NOT guarantee that the messages will be added in the
+//      order in which the functions were called. However, for any particular
+//      thread MMQ DOES GUARANTEE the order of the messages for the subsequent
+//      requests to push the message.
+//  This behavior of MMQ can be controlled via `DROP_ON_OVERFLOW` template
+//  argument.
 
 #include <condition_variable>
 #include <mutex>
 #include <string>
 #include <thread>
 #include <vector>
+#include <iostream>
+
+template <typename T_CONSUMER, typename T_MESSAGE>
+constexpr bool HasSimpleOnMessage(char) {
+  return false;
+}
+
+template <typename T_CONSUMER, typename T_MESSAGE>
+constexpr auto HasSimpleOnMessage(int)
+    -> decltype(std::declval<T_CONSUMER>().OnMessage(std::move(std::declval<T_MESSAGE>())), bool()) {
+  return true;
+}
+
+template <typename T_CONSUMER, typename T_MESSAGE>
+constexpr bool HasExtendedOnMessage(char) {
+  return false;
+}
+
+template <typename T_CONSUMER, typename T_MESSAGE>
+constexpr auto HasExtendedOnMessage(int)
+    -> decltype(std::declval<T_CONSUMER>().OnMessage(std::move(std::declval<T_MESSAGE>()), 0u), bool()) {
+  return true;
+}
+
+template <typename T_CONSUMER, typename T_MESSAGE, bool HAS_SIMPLE_ON_MESSAGE>
+struct ExportMessageImpl {};
+
+template <typename T_CONSUMER, typename T_MESSAGE>
+struct ExportMessageImpl<T_CONSUMER, T_MESSAGE, true> {
+  static void OnMessage(T_CONSUMER& consumer, T_MESSAGE&& message, size_t) {
+    consumer.OnMessage(std::move(message));
+  }
+};
+
+template <typename T_CONSUMER, typename T_MESSAGE>
+struct ExportMessageImpl<T_CONSUMER, T_MESSAGE, false> {
+  static void OnMessage(T_CONSUMER& consumer, T_MESSAGE&& message, size_t dropped_count) {
+    consumer.OnMessage(std::move(message), dropped_count);
+  }
+};
+
+template <typename T_CONSUMER, typename T_MESSAGE>
+void ExportMessage(T_CONSUMER& consumer, T_MESSAGE&& message, size_t dropped_count) {
+  static_assert(
+      !(HasSimpleOnMessage<T_CONSUMER, T_MESSAGE>(0) && HasExtendedOnMessage<T_CONSUMER, T_MESSAGE>(0)),
+      "Two versions of `OnMessage()` are defined, choose one.");
+  ExportMessageImpl<T_CONSUMER, T_MESSAGE, HasSimpleOnMessage<T_CONSUMER, T_MESSAGE>(0)>::OnMessage(
+      consumer, std::move(message), dropped_count);
+}
 
-template <typename CONSUMER, typename MESSAGE = std::string, size_t DEFAULT_BUFFER_SIZE = 1024>
+template <typename CONSUMER,
+          typename MESSAGE = std::string,
+          size_t DEFAULT_BUFFER_SIZE = 1024,
+          bool DROP_ON_OVERFLOW = true>
 class MMQ final {
  public:
   // Type of entries to store, defaults to `std::string`.
   typedef MESSAGE T_MESSAGE;
 
   // Type of the processor of the entries.
-  // It should expose one method, void OnMessage(T_MESSAGE&, size_t number_of_dropped_messages_if_any);
+  // It should expose one of the methods, accepting message by const or rvalue reference:
+  //
+  //   void OnMessage(T_MESSAGE&[&], size_t number_of_dropped_messages_if_any);
```

Comment:
I know it looks redundant, but `[const] ` would also fit the prototype. Why not show the `T_MESSAGE&&` syntax, and on the next line put `(Alternatively, void OnMessage(const T_MESSAGE&, size_t)` ?